### PR TITLE
[W11B01] Add tests for first 4 subtasks

### PR DESCRIPTION
**Pull request by julian-hartl**
I added basic tests for the first 4 subtasks. Feel free to add more tests and add tests for the last two subtasks.

### CODE REVIEW

**Comment by Anatoly03** [comment]
I am not passing Artemis Tests for the`generateResidualGraph ` method, but I'm passing your one test, ideas for edge cases?

File: w11b01/test/pgdp/pengusurvivors/ResidualGraphTest.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+                                """
+                        , "a", "a", List.of(List.of("a"))
+                ),
+                Arguments.of(
+                        """
+                                a -> b (1/1);
+                                b -> (0/2);
+                                """,
+                        "a", "c", List.of()
+                )
```
Comment (felixhauptmann) [inline w11b01/test/pgdp/pengusurvivors/ResidualGraphTest.java:117]: Add testcase with potential loop

```suggestion
                ),
                Arguments.of( // tests with potential loops
                        """
                                s -> a;
                                a -> b;
                                a -> t;
                                b -> c;
                                c -> a;
                                b -> t;
                                c -> t;
                                """,
                        "s", "t", List.of(List.of("s", "a", "b", "t"), List.of("s", "a", "t"), List.of("s", "a", "b", "t"), List.of("s", "a", "b", "c", "t"))
                )
```

File: w11b01/test/pgdp/pengusurvivors/ResidualGraphTest.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    static Stream<Arguments> findPathArguments() {
+        return Stream.of(
+                Arguments.of("""
+                        s -> a (1/1);
+                        s -> b (0/1);
+                        a -> c (1/1);
+                        b -> c (0/1);
+                        """, "s", "c", List.of(List.of("s", "b", "c"))),
```
Comment (felixhauptmann) [inline w11b01/test/pgdp/pengusurvivors/ResidualGraphTest.java:82]: `s -> a -> c` should also be valid, right?

```suggestion
                        """, "s", "c", List.of(List.of("s", "b", "c"), List.of("s", "a", "c"))),
```

File: w11b01/test/pgdp/pengusurvivors/ResidualGraphTest.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+    }
+
+    static Stream<Arguments> findPathArguments() {
+        return Stream.of(
+                Arguments.of("""
+                        s -> a (1/1);
+                        s -> b (0/1);
+                        a -> c (1/1);
+                        b -> c (0/1);
```
Comment (felixhauptmann) [inline w11b01/test/pgdp/pengusurvivors/ResidualGraphTest.java:81]: why do you specify the flow rate/capacity? Flowrate does not matter with this test and capacity is 1 per default anyway.

```suggestion
                        s -> a;
                        s -> b;
                        a -> c;
                        b -> c;
```

File: w11b01/test/pgdp/pengusurvivors/ResidualGraphTest.java (hunk 1 above)
Comment (felixhauptmann) [inline w11b01/test/pgdp/pengusurvivors/ResidualGraphTest.java:117]: This case does not really make sense since it is given that the graphs are always connected. This also produces invalid Graphviz output.

A proper alternative to this testcase would be a graph with edges where `capacity <= 0`:

```suggestion
                Arguments.of(
                        """
                                a -> b (0/1);
                                b -> c (0/0);
                                """,
                        "a", "c", List.of()
                )
```

File: w11b01/test/pgdp/pengusurvivors/ResidualGraphTest.java [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+                verticeString,
+                source,
+                sink
+        );
+        var graph = graphResult.graph();
+        GraphUtils.printGraph(graph);
+        var vertices = graphResult.vertices();
+        graph.generateResidualGraph();
+        var path = graph.findPathInResidual();
+        var matchingPaths = 0;
```
Comment (felixhauptmann) [inline w11b01/test/pgdp/pengusurvivors/ResidualGraphTest.java:55]: It would be nice to always output the generated solution.

```suggestion
        var path = graph.findPathInResidual();

        System.out.println("Your path was: " + (path == null ? "None found" : GraphUtils.stringifyPath(path)));

        var matchingPaths = 0;
```

File: w11b01/test/pgdp/pengusurvivors/ResidualGraphTest.java [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        for (var expectedPath : expectedPaths) {
+            if (path == null || expectedPath.size() != path.size()) continue;
+            for (int i = 0; i < path.size(); i++) {
+                var expectedVertex = vertices.get(expectedPath.get(i));
+                var actualVertex = path.get(i);
+                if (expectedVertex != actualVertex) continue pathloop;
+            }
+            matchingPaths++;
+        }
+        assertEquals(1, matchingPaths, "No matching path was found. Your path was: " + (path == null ? "None found" : GraphUtils.stringifyPath(path)));
```
Comment (felixhauptmann) [inline w11b01/test/pgdp/pengusurvivors/ResidualGraphTest.java:71]: This would be redundant if you accept my [other suggestion](https://github.com/MaximilianAnzinger/pgdp2223-tests/pull/273/files#diff-bdd3b743ffcf3af44228237f752bfe8d64f63eeb2f8d408339929a2e60879446R54-R55).

```suggestion
        assertEquals(1, matchingPaths, "No matching path was found.");
```

**Review by felixhauptmann — CHANGES_REQUESTED** [review]

File: w11b01/test/pgdp/pengusurvivors/ResidualGraphTest.java [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+                                c -> t;
+                                """,
+                        "s", "t", List.of(List.of("s", "a", "b", "t"), List.of("s", "a", "c", "t"), List.of("s", "a", "b", "c", "t"))
+                ),
+                Arguments.of(
+                        """
+                                a -> a;
+                                """
+                        , "a", "a", List.of(List.of("a"))
+                ),
```
Comment (felixhauptmann) [inline w11b01/test/pgdp/pengusurvivors/ResidualGraphTest.java:110]: I think this test should get removed as it is clearly stated that there are no loops allowed. Therefore there must not exist any solutions with `path.size() == 1` because this would indicate a self-loop.

Apart from that... what would be the augmenting flow of such path.

```suggestion
```

**Review by felixhauptmann — CHANGES_REQUESTED** [review]

File: w11b01/test/pgdp/pengusurvivors/ResidualGraphTest.java [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+    static Stream<Arguments> getMaxFlowArguments() {
+        return Stream.of(
+                Arguments.of(
+                        """
+                                a -> b (4/10);
+                                b -> c (2/6);
+                                """,
+                        "a", "c",
+                        6
```
Comment (felixhauptmann) [inline w11b01/test/pgdp/pengusurvivors/ResidualGraphTest.java:172]: The expected flowrate should be 8.

```suggestion
                        8
```

**Review by felixhauptmann — CHANGES_REQUESTED** [review]

File: w11b01/test/pgdp/pengusurvivors/ResidualGraphTest.java (hunk 3 above)
Comment (julian-hartl) [reply w11b01/test/pgdp/pengusurvivors/ResidualGraphTest.java:81]: I guess that's just left over from some earlier versions of these tests.

**Comment by julian-hartl** [comment]
@felixhauptmann Thank you very much for those helpful suggestions :) Accepted all of them and if there is nothing left to do, I'd say @nilsreichardt you know what to do.
